pkg/controller/machineset/controller.go

```diff
@@ -182,7 +204,218 @@ func (c *MachineSetControllerImpl) adoptOrphan(machineSet *v1alpha1.MachineSet,
+			glog.V(4).Infof("creating a machine ( spec.replicas(%d) > currentMachineCount(%d) )", *(ms.Spec.Replicas), len(filteredMachines))
+			machine := c.createMachine(ms)
+			_, err := c.clusterAPIClient.ClusterV1alpha1().Machines(ms.Namespace).Create(machine)
+			if err != nil {
+				glog.Errorf("unable to create a machine = %s, due to %v", machine.Name, err)
+				errstrings = append(errstrings, err.Error())
+			}
+		}
+
+		if errstrings != nil {
+			return fmt.Errorf(strings.Join(errstrings, "; "))
+		}
+
+		return nil
+	} else if diff > 0 {
+		glog.V(4).Infof("Too many replicas for %v %s/%s, need %d, deleting %d", controllerKind, ms.Namespace, ms.Name, *(ms.Spec.Replicas), diff)
+
+		// Choose which Machines to delete.
+		machinesToDelete := getMachinesToDelete(filteredMachines, diff)
+
+		errCh := make(chan error, diff)
+		var wg sync.WaitGroup
+		wg.Add(diff)
+		for _, machine := range machinesToDelete {
+			go func(targetMachine *v1alpha1.Machine) {
+				defer wg.Done()
+				err := c.clusterAPIClient.ClusterV1alpha1().Machines(ms.Namespace).Delete(targetMachine.Name, &metav1.DeleteOptions{})
+				if err != nil {
+					glog.Errorf("unable to delete a machine = %s, due to %v", machine.Name, err)
+					errCh <- err
+				}
+			}(machine)
+		}
+		wg.Wait()
+
+		select {
+		case err := <-errCh:
+			// all errors have been reported before and they're likely to be the same, so we'll only return the first one we hit.
+			if err != nil {
+				return err
+			}
+		default:
+		}
+	}
+
+	return nil
+}
+
+func (c *MachineSetControllerImpl) reconcileAfter(machineSet *v1alpha1.MachineSet, after time.Duration) {
+	timer := time.NewTimer(after)
+	go func(newMS v1alpha1.MachineSet) {
+		<-timer.C
+		c.Reconcile(&newMS)
+	}(*machineSet)
+}
+
+// updateMachineSetStatus attempts to update the Status.Replicas of the given MachineSet, with a single GET/PUT retry.
+func updateMachineSetStatus(c machinesetclientset.MachineSetInterface, ms *v1alpha1.MachineSet, newStatus v1alpha1.MachineSetStatus) (*v1alpha1.MachineSet, error) {
+	// This is the steady state. It happens when the MachineSet doesn't have any expectations, since
+	// we do a periodic relist every 30s. If the generations differ but the replicas are
+	// the same, a caller might've resized to the same replica count.
+	if ms.Status.Replicas == newStatus.Replicas &&
+		ms.Status.FullyLabeledReplicas == newStatus.FullyLabeledReplicas &&
+		ms.Status.ReadyReplicas == newStatus.ReadyReplicas &&
+		ms.Status.AvailableReplicas == newStatus.AvailableReplicas &&
+		ms.Generation == ms.Status.ObservedGeneration {
+		return ms, nil
+	}
+
+	// Save the generation number we acted on, otherwise we might wrongfully indicate
+	// that we've seen a spec update when we retry.
+	// TODO: This can clobber an update if we allow multiple agents to write to the
+	// same status.
+	newStatus.ObservedGeneration = ms.Generation
+
+	var getErr, updateErr error
+	var updatedMS *v1alpha1.MachineSet
+	for i, ms := 0, ms; ; i++ {
```

Comment:
This looks interesting. What does the additional ms:=0, ms do?

Reply:
this was copied from replicaset controller, which now that you point out, isnt really needed.
after checking with krousey, this sets i to 0, and ms to ms, multi-variable initialization.
but since ms is already a variable, it doesnt do anything.
fixed it to be a simple i while loop.